File: js/src/mldsa/decompose.js
/**
 * Decomposition functions for ML-DSA (FIPS 204)
 * power2Round, decompose, highBits, lowBits, makeHint, useHint
 */

import { Q, modQ } from './field.js';

/**
 * Power2Round: decompose r into (r1, r0) such that r = r1 * 2^d + r0
 * where d = 13 and r0 is in (-2^(d-1), 2^(d-1)].
 * @param {number} r - value in [0, Q)
 * @returns {[number, number]} [r1, r0]
 */
export function power2Round(r) {
  const d = 13;
  const rPlus = modQ(r);
  const r0 = centeredMod(rPlus, 1 << d);
  return [((rPlus - r0) >>> d), r0];
}

/**
 * Centered modular reduction: returns r mod m in range (-m/2, m/2].
 * @param {number} r
 * @param {number} m
 * @returns {number}
 */
function centeredMod(r, m) {
  let r0 = r % m;
  if (r0 < 0) r0 += m;
  if (r0 > (m >>> 1)) r0 -= m;
  return r0;
}

/**
 * Decompose r into (r1, r0) such that r = r1 * alpha + r0
 * with r0 in (-alpha/2, alpha/2] except for corner case.
 * alpha = 2*gamma2.
 * @param {number} r - value in [0, Q)
 * @param {number} alpha - decomposition base (2*gamma2)
 * @returns {[number, number]} [r1, r0]
 */
export function decompose(r, alpha) {
  const rPlus = modQ(r);
  let r0 = centeredMod(rPlus, alpha);
  let r1;
  if (rPlus - r0 === Q - 1) {
    r1 = 0;
    r0 = r0 - 1;
  } else {
    r1 = (rPlus - r0) / alpha;
  }
  return [r1, r0];
}

/**
 * Extract high bits of r with respect to alpha = 2*gamma2.
 * @param {number} r
 * @param {number} alpha
 * @returns {number}
 */
export function highBits(r, alpha) {
  return decompose(r, alpha)[0];
}

/**
 * Extract low bits of r with respect to alpha = 2*gamma2.
 * @param {number} r
 * @param {number} alpha
 * @returns {number}
 */
export function lowBits(r, alpha) {
  return decompose(r, alpha)[1];
}

/**
 * Compute a hint bit: 1 if highBits of z differs from highBits of r.
 * @param {number} z - low part
 * @param {number} r - full value
 * @param {number} alpha
 * @returns {number} 0 or 1
 */
export function makeHint(z, r, alpha) {
  const r1 = highBits(r, alpha);
  const v1 = highBits(modQ(r + z), alpha);
  return r1 !== v1 ? 1 : 0;
}

/**
 * Use a hint to recover the correct high bits.
 * @param {number} hint - 0 or 1
 * @param {number} r - value in [0, Q)
 * @param {number} alpha
 * @returns {number} corrected high bits
 */
export function useHint(hint, r, alpha) {
  const [r1, r0] = decompose(r, alpha);
  if (hint === 0) {
    return r1;
  }
  const m = (Q - 1) / alpha;
  if (r0 > 0) {
    return (r1 + 1) % m;
  } else {
    return (r1 - 1 + m) % m;
  }
}

File: rust/ml-dsa/src/decompose.rs
//! Decomposition, rounding, and hint functions for ML-DSA (FIPS 204).
//!
//! These functions split field elements into high and low parts for the
//! signature scheme's rounding and hint mechanisms.

use crate::field::Q;
use crate::params::D;

/// Split r into (r1, r0) such that r = r1 * 2^d + r0, with r0 in [-(2^(d-1)-1), 2^(d-1)].
///
/// d = 13 for ML-DSA.
/// Returns (r1, r0) where r0 is represented mod q.
pub fn power2_round(r: u32) -> (u32, u32) {
    // r+ = r mod q (already assumed)
    let half = 1u32 << (D - 1); // 2^12 = 4096
    let r0_raw = r & ((1 << D) - 1); // r mod 2^d

    let (r1, r0) = if r0_raw > half {
        // r0 = r0_raw - 2^d (negative), r1 = (r + (2^d - r0_raw)) / 2^d
        ((r + (1 << D) - r0_raw) >> D, Q - ((1 << D) - r0_raw))
    } else {
        // r0 = r0_raw (non-negative)
        ((r - r0_raw) >> D, r0_raw)
    };
    (r1, r0)
}

/// Decompose r into (r1, r0) such that r ≡ r1 * alpha + r0, with
/// |r0| <= alpha/2. alpha = 2 * gamma2.
///
/// Returns (r1, r0) where r0 is represented mod q.
pub fn decompose(r: u32, alpha: u32) -> (u32, u32) {
    // r+ = r mod+ q  (already in [0, q))
    decompose_spec(r, alpha)
}

/// Decompose following FIPS 204 Algorithm 36 precisely.
fn decompose_spec(r: u32, alpha: u32) -> (u32, u32) {
    // Step 1: r0 = r mod+ alpha (centered modular reduction)
    let r0_pos = r % alpha;
    let r0_centered: i32;
    if r0_pos > alpha / 2 {
        r0_centered = r0_pos as i32 - alpha as i32; // negative
    } else {
        r0_centered = r0_pos as i32;
    }

    // Step 2: if r - r0 == q - 1
    let r_minus_r0 = if r0_centered >= 0 {
        r - r0_centered as u32
    } else {
        r + (-r0_centered) as u32
    };

    let (r1, r0_final);
    if r_minus_r0 == Q - 1 {
        r1 = 0u32;
        r0_final = if r0_centered > 0 {
            (r0_centered - 1) as u32
        } else if r0_centered == 0 {
            Q - 1 // -1 mod q
        } else {
            // r0_centered - 1 is more negative
            let val = r0_centered - 1;
            ((val as i64 + Q as i64) % Q as i64) as u32
        };
    } else {
        r1 = r_minus_r0 / alpha;
        r0_final = if r0_centered >= 0 {
            r0_centered as u32
        } else {
            ((r0_centered as i64) + Q as i64) as u32
        };
    }

    (r1, r0_final)
}

/// Return the high bits of r with respect to alpha = 2 * gamma2.
#[inline]
pub fn high_bits(r: u32, alpha: u32) -> u32 {
    decompose(r, alpha).0
}

/// Return the low bits of r with respect to alpha = 2 * gamma2.
#[inline]
pub fn low_bits(r: u32, alpha: u32) -> u32 {
    decompose(r, alpha).1
}

/// Compute the hint bit: returns true iff high_bits(r) != high_bits(r + z mod q).
pub fn make_hint(z: u32, r: u32, alpha: u32) -> bool {
    let r1 = high_bits(r, alpha);
    let rz = crate::field::field_add(r, z);
    let v1 = high_bits(rz, alpha);
    r1 != v1
}

/// Recover the correct high bits using a hint.
///
/// If h is false, return high_bits(r). If h is true, adjust by ±1
/// depending on the sign of low_bits(r).
pub fn use_hint(hint: bool, r: u32, alpha: u32) -> u32 {
    let (r1, r0) = decompose(r, alpha);
    if !hint {
        return r1;
    }

    // r0 > 0 means we round up, r0 <= 0 means round down
    // r0 is in centered form: positive means r0 < alpha/2, stored as is
    // negative means r0 stored as Q - |r0|
    let m = (Q - 1) / alpha; // number of possible high-bits values

    let r0_positive = r0 <= alpha / 2; // r0 >= 0 in centered rep

    if r0_positive {
        // r0 > 0: return (r1 + 1) mod m
        if r1 + 1 >= m {
            0
        } else {
            r1 + 1
        }
    } else {
        // r0 <= 0: return (r1 - 1) mod m
        if r1 == 0 {
            m - 1
        } else {
            r1 - 1
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_power2_round_identity() {
        // For various r: r1 * 2^d + r0 ≡ r (mod q)
        let test_values = [0u32, 1, 100, 4096, 8191, 8192, Q / 2, Q - 1];
        for &r in &test_values {
            let (r1, r0) = power2_round(r);
            // r1 * 2^d + r0 mod q = r
            let reconstructed = crate::field::field_add(
                crate::field::field_mul(r1, 1 << D),
                r0,
            );
            assert_eq!(
                reconstructed, r,
                "power2_round identity failed for r={}",
                r
            );
        }
    }

    #[test]
    fn test_power2_round_r0_range() {
        // r0 should be in [-(2^(d-1)-1), 2^(d-1)]
        let half = 1u32 << (D - 1);
        for r in (0..Q).step_by(1000) {
            let (_, r0) = power2_round(r);
            // r0 is either in [0, half] or in [Q - half + 1, Q - 1]
            assert!(
                r0 <= half || r0 >= Q - half + 1,
                "r0={} out of centered range for r={}",
                r,
                r
            );
        }
    }

    #[test]
    fn test_decompose_identity_alpha_190464() {
        let alpha = 190464u32; // 2 * gamma2 for ML-DSA-44
        for r in (0..Q).step_by(10000) {
            let (r1, r0) = decompose(r, alpha);
            let reconstructed = crate::field::field_add(
                crate::field::field_mul(r1, alpha),
                r0,
            );
            assert_eq!(
                reconstructed, r,
                "decompose identity failed for r={}, got r1={}, r0={}",
                r, r1, r0
            );
        }
    }

    #[test]
    fn test_decompose_identity_alpha_523776() {
        let alpha = 523776u32; // 2 * gamma2 for ML-DSA-65/87
        for r in (0..Q).step_by(10000) {
            let (r1, r0) = decompose(r, alpha);
            let reconstructed = crate::field::field_add(
                crate::field::field_mul(r1, alpha),
                r0,
            );
            assert_eq!(
                reconstructed, r,
                "decompose identity failed for r={}, got r1={}, r0={}",
                r, r1, r0
            );
        }
    }

    #[test]
    fn test_high_low_bits_consistency() {
        let alpha = 523776u32;
        for r in (0..Q).step_by(50000) {
            let h = high_bits(r, alpha);
            let l = low_bits(r, alpha);
            let (r1, r0) = decompose(r, alpha);
            assert_eq!(h, r1);
            assert_eq!(l, r0);
        }
    }

    #[test]
    fn test_make_hint_use_hint_roundtrip() {
        let alpha = 190464u32;
        // If hint = make_hint(z, r), then use_hint(hint, r+z) should give
        // the same as high_bits(r + z) when correctly applied.
        for r in (0..Q).step_by(100000) {
            for z in (0..Q).step_by(200000) {
                let rz = crate::field::field_add(r, z);
                let hint = make_hint(z, r, alpha);
                let _recovered = use_hint(hint, rz, alpha);
                // use_hint(make_hint(z, r), r + z) should = high_bits(r + z)
                // Actually per spec: use_hint applied to (hint, r) gives high_bits(r+z)
                // But we need to verify the spec's property:
                // high_bits(r) should be recoverable
            }
        }
    }

    #[test]
    fn test_hint_basic() {
        let alpha = 190464u32;
        // When z = 0, hint should be false (no change)
        for r in (0..Q).step_by(100000) {
            let hint = make_hint(0, r, alpha);
            assert!(!hint, "hint should be false when z=0, r={}", r);
        }
    }

    #[test]
    fn test_use_hint_no_change() {
        let alpha = 523776u32;
        for r in (0..Q).step_by(100000) {
            let r1 = high_bits(r, alpha);
            let recovered = use_hint(false, r, alpha);
            assert_eq!(recovered, r1);
        }
    }
}
